server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1072,6 +1077,28 @@ private void fillSegmentInfo(SegmentReader segmentReader, boolean search, Map<St
         }
         segment.segmentSort = info.info.getIndexSort();
         segment.attributes = info.info.getAttributes();
+        Codec codec = info.info.getCodec();
+        segment.codec = codec.getName();
+        if (codec instanceof PerFieldMapperCodec) {
+            try {
+                // potentially expensive with many fields
+                FieldInfos fieldInfos = segmentReader.getFieldInfos();
+                if (fieldInfos.hasVectors()) {
+                    for (FieldInfo fieldInfo : fieldInfos) {
+                        String name = fieldInfo.getName();
+                        if (fieldInfo.getVectorDimension() > 0) {
+                            KnnVectorsFormat knnVectorsFormatForField = ((PerFieldMapperCodec) codec).getKnnVectorsFormatForField(name);
+                            if (segment.knnFormats == null) {
+                                segment.knnFormats = new HashMap<>();
+                            }
+                            segment.knnFormats.put(name, knnVectorsFormatForField.getName());
+                        }
+                    }
+                }
+            } catch (AlreadyClosedException ace) {
+                // silently ignore ?
+            }
```

Comment:
`SegmentReader#getFieldInfos` calls `#ensureOpen` which might rise (unchecked) `AlreadyClosedException`, I think it'd be safer to silently fail.